### PR TITLE
Also pass yargs to the fail handler

### PR DESCRIPTION
**Pull request by mkawalec**
This is needed to display a complete help message for a command that failed in a top-level `fail` handler

### CODE REVIEW

**Comment by bcoe** [comment]
@mkawalec love it. 👍

**Comment by mkawalec** [comment]
@bcoe: Should I change something, or can we merge this? :)

**Comment by bcoe** [comment]
@mkawalec sorry for the slow turnaround! looks great, but could I perhaps get you to add a test? Partially because it would be neat to see how you plan on using this functionality 👍

**Comment by mkawalec** [comment]
I'll add all of that early next week, thanks for the review.

**Comment by bcoe** [comment]
@mkawalec would love to land this, any news on that test :)

**Comment by mkawalec** [comment]
@bcoe: I've added the test, but it fails. The reason for that is that the exception thrown in command handler is not caught and so `fail` is never called. However, in a program I have that uses yargs the `fail` handler is called if an exception is thrown in the command handler. Is this a problem with the test harness or my mistake somewhere?

**Comment by bcoe** [comment]
@mkawalec when I have some time this weekend I'll take a look at the harness 👍

**Comment by bcoe** [comment]
@mkawalec _this weekend_ has quickly become _this following weekend_, for which I apologize; I haven't forgotten about this pull.

**Comment by mkawalec** [comment]
No worries @bcoe, I know the pain well :)

File: README.md [hunk 1]
```diff
@@ -1055,15 +1055,17 @@ yargs have been validated.
 
 Method to execute when a failure occurs, rather than printing the failure message.
 
-`fn` is called with the failure message that would have been printed and the
-`Error` instance originally thrown, if any.
+`fn` is called with the failure message that would have been printed, the
+`Error` instance originally thrown and yargs state when the failure
```
Comment (bcoe) [inline README.md]: slight editing nit, let's change this to:

`fn` is called with the failure message that would have been printed, the
`Error` instance originally thrown, and the yargs instance when the failure occurred.

File: test/usage.js [hunk 2]
```diff
@@ -473,6 +473,31 @@ describe('usage tests', function () {
         r.logs.should.deep.equal(['foo'])
         r.should.have.property('exit').and.be.false
       })
+
+      it('gives the ability to log a per-command error message if failure occurs in a command', function () {
```
Comment (bcoe) [inline test/usage.js]: I finally had the time to debug this test for you, here's the code I pulled together:

```js
      it('is invoked with yargs instance as third argument', function () {
        var r = checkUsage(function () {
          return yargs('foo')
            .command('foo', 'desc', {
              bar: {
                describe: 'bar command'
              }
            }, function (argv) {
              throw new Error('blah')
            })
            .fail(function (message, error, yargs) {
              yargs.showHelp()
            })
            .exitProcess(false)
            .wrap(null)
            .argv
        })

        r.errors[0].should.contain('bar command')
      })
```

Comment (bcoe) [inline test/usage.js]: @mkawalec awesome 👍

**Comment by mkawalec** [comment]
@bcoe please rereview :)

**Pull request merged**